apps/labs/posts/numpy-string-ufuncs.md

```diff
@@ -0,0 +1,546 @@
+
+## Next up was `add`
+
+Next on our list of functions to implement was `add`. This was not a new ufunc
+as `np.add` already existed for other dtypes. All we had to do was just implement
+the new loop (which was essentially the same as `isalpha` with just one more input)
+and everything would be fine, right? Wrong! As soon as I built NumPy and ran a
+Python REPL to test, I was met with the following error:
+
+```python3
+>>> import numpy as np
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+  File "/numpy/build-install/usr/lib/python3.11/site-packages/numpy/__init__.py", line 114, in <module>
+    from numpy.__config__ import show as show_config
+  File "/numpy/build-install/usr/lib/python3.11/site-packages/numpy/__config__.py", line 4, in <module>
+    from numpy._core._multiarray_umath import (
+  File "/numpy/build-install/usr/lib/python3.11/site-packages/numpy/_core/__init__.py", line 23, in <module>
+    from . import multiarray
+  File "/numpy/build-install/usr/lib/python3.11/site-packages/numpy/_core/multiarray.py", line 10, in <module>
+    from . import overrides
+  File "/numpy/build-install/usr/lib/python3.11/site-packages/numpy/_core/overrides.py", line 8, in <module>
+    from numpy._core._multiarray_umath import (
+TypeError: must provide a `resolve_descriptors` function if any output DType is parametric. (method: templated_string_add)
+```
+
+The important part of this traceback is the error message: must provide a
+`resolve_descriptors` function if any output DType is parametric. That is a very
+fancy way to say that NumPy does not know enough about the output dtype in order
+to be able to register the loop. Remember that `isalpha` returns booleans. That's
+easy enough for NumPy to do, since booleans are fixed-size dtypes. However, string
+dtypes, both unicode and bytes, are parametric dtypes, meaning that their size
+is dependent on the array items, specifically their lengths. In order for NumPy
+to be able to register the loop, it needs a `resolve_descriptors` function which
+specifies the parameters for all the necessary dtypes. This function has the
+following signature:
+
+```cpp
+NPY_CASTING resolve_descriptors(struct PyArrayMethodObject_tag *method,
+                                PyArray_DTypeMeta **dtypes,
+                                PyArray_Descr **given_descrs,
+                                PyArray_Descr **loop_descrs,
+                                npy_intp *view_offset)
+```
+
+The important arguments here are `given_descrs` and `loop_descrs`:
+- `given_descrs` is an array of input and output (if any were given
+   by the user) descriptor instances. A descriptor instance is a descriptor
+   for an array, and includes information about the size of each item.
+- `loop_descrs` is an array of descriptors that must be filled in by the resolve
+   descriptors implementation. These will be the descriptors that will be used
+   in order to call the loop with the correct information.
```

Comment:
maybe also link to the numpy 2.0 ArrayMethod docs which go over all this

Reply:
https://numpy.org/devdocs/reference/c-api/array.html#c.NPY_METH_resolve_descriptors (not yet available in a build of the docs for a stable release)